pkg/validation/internal/operatorhub.go

```diff
@@ -74,111 +96,303 @@ func validateBundleOperatorHub(bundle *manifests.Bundle) errors.ManifestResult {
-			if err != nil {
-				errs = append(errs, fmt.Errorf("csv.Spec.Links url %s is invalid: %v", link.URL, err))
+	}
+
+	// if the k8s value was informed and it is >=1.16 we should check
+	// if the k8s value was not informed we also should check since the
+	// check should occurs with any minKubeVersion value:
+	// - if minKubeVersion empty then means that the project can be installed in any version
+	// - if minKubeVersion any version defined it means that we are considering install
+	// in any upper version from that where the check is always applied
+	if !isVersionProvided || semVerVersionProvided.GE(semVerk8sVerV1betav1Deprecated) {
+		if len(bundle.V1beta1CRDs) > 0 {
+			var crdApiNames []string
+			for _, obj := range bundle.V1beta1CRDs {
+				crdApiNames = append(crdApiNames, obj.Name)
+			}
+
+			// isUnsupported is true only if the key/value OR minKubeVersion were informed and are >= 1.22
+			isUnsupported := semVerVersionProvided.GE(semVerK8sVerV1betav1Unsupported) ||
+				semverMinKube.GE(semVerK8sVerV1betav1Unsupported)
+			// We only raise an error when the version >= 1.22 was informed via
+			// the k8s key/value option or is specifically defined in the CSV
+			if isUnsupported {
+				errs = append(errs, fmt.Errorf("%s: %+q should be migrated", crdv1beta1DeprecationMsg, crdApiNames))
+			} else {
+				warns = append(warns, fmt.Errorf("%s: %+q should be migrated", crdv1beta1DeprecationMsg, crdApiNames))
 			}
 		}
 	}
 
-	if csv.GetAnnotations() == nil {
-		csv.SetAnnotations(make(map[string]string))
-	}
+	return errs, warns
+}
 
-	if capability, ok := csv.ObjectMeta.Annotations["capabilities"]; ok {
-		if _, ok := validCapabilities[capability]; !ok {
-			errs = append(errs, fmt.Errorf("csv.Metadata.Annotations.Capabilities %s is not a valid capabilities level", capability))
+// validateHubCSVSpec will check the CSV against the criteria to publish an
+// operator bundle in the OperatorHub.io
+func validateHubCSVSpec(csv v1alpha1.ClusterServiceVersion) CSVChecks {
+	checks := CSVChecks{csv: csv, errs: []error{}, warns: []error{}}
+
+	checks = checkSpecProviderName(checks)
+	checks = checkSpecMaintainers(checks)
+	checks = checkSpecLinks(checks)
+	checks = checkAnnotations(checks)
+	checks = checkSpecVersion(checks)
+	checks = checkSpecIcon(checks)
+	checks = checkSpecMinKubeVersion(checks)
+	checks = checkBundleName(checks)
+
+	return checks
+}
+
+type CSVChecks struct {
+	csv   v1alpha1.ClusterServiceVersion
+	errs  []error
+	warns []error
+}
+
+// checkBundleName will validate the operator bundle name informed via CSV.metadata.name.
+// The motivation for the following check is to ensure that operators authors knows that operator bundles names should
+// follow a name and versioning convention
+func checkBundleName(checks CSVChecks) CSVChecks {
+
+	// Check if is following the semver
+	re := regexp.MustCompile("([0-9]+)\\.([0-9]+)\\.([0-9]+)(?:-([0-9A-Za-z-]+(?:\\.[0-9A-Za-z-]+)*))?(?:\\+[0-9A-Za-z-]+)?$")
```

Comment:
This regexp is incorrect since it allows multiple hyphens and weird garbage at the end of the version itself. How about splitting the string by `.` and parsing the last n-1 values with the semver lib after stripping “v” from the first of them? That should always give you a correct value and avoids complex regexp.

Reply:
Its goal is to get what the authors added as a version. I added a test to prove that the warning will work well if the name has multiple hyphens as described. See (memcached-operator.v1--1.0). Then, I would prefer to move forward here with the regex it shows easier to keep maintained and more assertive.

Reply:
If the "extra stuff" formatting is intentional, then I'll resolve this issue.